File: scripts/eip55_recompute.pl
#!/usr/bin/env perl
use strict; use warnings;
use Digest::Keccak qw(keccak_256_hex);

my $addr = shift @ARGV;
if (!defined $addr) {
  print STDERR "usage: eip55_recompute.pl 0x<hexaddr>\n";
  exit 2;
}

$addr =~ s/^0x//i;
my $lc = lc($addr);
my $h = keccak_256_hex($lc);
my $out = '';
for my $i (0..39) {
  my $c = substr($lc, $i, 1);
  my $d = hex(substr($h, $i, 1));
  if ($c =~ /[a-f]/ && $d >= 8) { $out .= uc($c) } else { $out .= $c }
}
print "0x$out\n";

File: scripts/eip55_recompute.py
#!/usr/bin/env python3
import sys

def checksum(addr: str) -> str:
    if addr.startswith("0x") or addr.startswith("0X"):
        addr = addr[2:]
    lc = addr.lower()
    from Crypto.Hash import keccak
    k = keccak.new(digest_bits=256)
    k.update(lc.encode("ascii"))
    h = k.hexdigest()
    out = []
    for i, c in enumerate(lc):
        if c in "abcdef" and int(h[i], 16) >= 8:
            out.append(c.upper())
        else:
            out.append(c)
    return "0x" + "".join(out)

def main():
    if len(sys.argv) != 2:
        print("usage: eip55_recompute.py 0x<hexaddr>", file=sys.stderr)
        sys.exit(2)
    try:
        print(checksum(sys.argv[1]))
    except Exception as e:
        print("error:", e, file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

